archive/100-theorems-list/90_stirlings_formula.lean

```diff
@@ -0,0 +1,874 @@
+      (1 / 4) * (1 / (k.succ * k.succ.succ)) :=
+     begin
+       intro k,
+       norm_cast,
+       field_simp,
+       simp only [one_div, inv_inj],
+       ring_nf,
+     end,
+    refine sum_congr rfl (λ k, λ hk, hi k),
+   end
+   ... = 1 / 4 * ∑ k in range n, 1 / (k.succ * k.succ.succ) : by rw mul_sum
+   ... ≤ 1 / 4 * 1 :
+   begin
+     refine (mul_le_mul_left _).mpr _,
+     { exact div_pos one_pos four_pos },
+     { convert rat.cast_le.mpr (partial_sum_consecutive_reciprocals n),
+       rw rat_cast_sum,
+       push_cast,
+       exact rat.cast_one.symm },
+   end
+   ... = 1 / 4 : by rw mul_one,
+end
+
+/-- The sequence `bn` is bounded below by `3/4 - 1/2 * log 2`  for `n ≥ 1`. -/
+lemma bn_bounded_by_constant : ∀ (n : ℕ), 3 / (4 : ℝ) - 1 / 2 * log 2 ≤ bn n.succ :=
+begin
+  intro n,
+  calc
+  bn n.succ ≥ bn 1 - 1 / 4 : sub_le.mp (bn_bounded_aux n)
+   ... = (log ((1 : ℕ).factorial) - 1 / 2 * log (2 * (1 : ℕ)) - (1 : ℕ) *
+          log ((1 : ℕ) / (exp 1))) - 1 / 4 : by rw bn_formula 0
+   ... = 0 - 1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 :
+      by simp only [factorial_one, cast_one, log_one, one_div, mul_one, log_inv, log_exp, mul_neg]
+   ... = -1 / 2 * log 2 - log (1 / (exp 1)) - 1 / 4 : by ring
+   ... = -1 / 2 * log 2 + 1 - 1 / 4  : by simp only [one_div, log_inv, log_exp, sub_neg_eq_add]
+   ... = -1 / 2 * log 2 + 3 / 4      : by ring
+   ... = 3 / (4 : ℝ) - 1 / 2 * log 2 : by ring,
+end
+
+/-- The sequence `bn` is bounded below. -/
+lemma bn_has_lower_bound : (lower_bounds (set.range (λ (k : ℕ), bn k.succ))).nonempty :=
+begin
+   use 3 / (4 : ℝ) - 1 / 2 * log 2,
+   intros,
+   rw lower_bounds,
+   rw [set.mem_set_of_eq],
+   simp only [set.mem_range, forall_apply_eq_imp_iff', forall_exists_index],
+   exact bn_bounded_by_constant,
+end
+
+/--
+Any sequence `bn` of real numbers that is monotone decreasing and bounded below has
+a limit in the real numbers.
+-/
+lemma monotone_convergence (bn : ℕ → ℝ) (h_sd : ∀ (n m : ℕ), n ≤ m → bn m ≤ bn n)
+  (h_bounded : (lower_bounds (set.range bn)).nonempty) : ∃ (m : ℝ), tendsto bn at_top (𝓝 m) :=
+begin
+ use Inf (set.range bn),
+ exact tendsto_at_top_is_glb h_sd (real.is_glb_Inf (set.range bn)
+   (set.range_nonempty bn) h_bounded),
+end
+
+/-- The sequence `an` is positive for `n > 0`  -/
+lemma an'_pos : ∀ (n : ℕ), 0 < an n.succ :=
+ (λ n, div_pos (cast_pos.mpr (factorial_pos n.succ))
+    (mul_pos ((real.sqrt_pos).mpr (mul_pos two_pos (cast_pos.mpr (succ_pos n))))
+    (pow_pos (div_pos (cast_pos.mpr (succ_pos n)) (exp_pos 1)) n.succ)))
+
+/--
+The sequence `an` has the explicit lower bound exp (3/4 - 1/2 * log 2)
+-/
+lemma an'_bounded_by_pos_constant : ∀ (n : ℕ), exp (3 / (4 : ℝ) - 1 / 2 * log 2) ≤ an n.succ :=
+begin
+  intro n,
+  rw ← (le_log_iff_exp_le (an'_pos n)),
+  exact bn_bounded_by_constant n,
+end
+
+/-- The sequence `an` is monotone decreasing -/
+lemma an'_antitone : ∀ (n m : ℕ), n ≤ m → an m.succ ≤ an n.succ :=
```

Comment:
You should directly prove `antitone an` instead.

Reply:
see comment above, I don't really need `antitone an`, but a weaker statement. Perhaps I could change the name instead?

Reply:
I changed the statement to `antitone (stirling_seq ∘ succ)`, which is definitionally equal. Also changed the documentation. The name was already changed.